File: backend/src/Services/BillTypeService.cs
using API.Entities;
using API.Interfaces.Repositories;
using API.Types;

namespace API.Services;

public class BillTypeService(IBillTypeRepository billTypeRepository) {

    public async Task<Page<BillType>> GetAll(string filter, int page, int limit) {

        Page<BillType> billTypes = await billTypeRepository.GetAll(filter, page, limit);

        return billTypes;

    }

    public async Task<BillType?> GetBillTypeById(int id) {

        BillType? billType = await billTypeRepository.GetById(id);

        return billType;

    }

    public async Task<BillType?> CreateBillType(string name) {

        BillType billType = new() {
            Name = name
        };

        try {
            return await billTypeRepository.Create(billType);
        } catch {
            return null;
        }

    }

    public async Task<BillType?> UpdateBillType(int id, string? name) {

        BillType? billType = await billTypeRepository.GetById(id);

        if(billType == null) {
            return null;
        }

        billType.Name = name ?? billType.Name;

        try {
            return await billTypeRepository.Update(billType);
        } catch {
            return null;
        }

    }

    public async Task<BillType?> DeleteBillType(int id) {

        BillType? billType = await billTypeRepository.GetById(id);

        if(billType == null) {
            return null;
        }

        try {
            return await billTypeRepository.Delete(billType);
        } catch {
            return null;
        }

    }

}

File: frontend/src/services/bill-type.service.ts
import { APIError } from "@/models/api-error.models";
import { BillType } from "@/models/bill-type.models";
import { Page } from "@/models/page";
import { Service } from "@/services/service";
import { optional } from "@/types/optional";

export class BillTypeService extends Service {

    static async getAllBillTypes(filter : string, page : number, limit : number) {

        const response = await this.axios.get<Page<BillType> | APIError>('/api/bill-type', {
            params: {
                filter,
                page,
                limit
            }
        });

        return response.data;

    }

    static async createBillType(name : string) {

        const response = await this.axios.post<BillType | APIError>('/api/bill-type', {
            name
        });

        return response.data;

    }

    static async updateBillType(id : number, name : optional<string>) {

        const response = await this.axios.put<BillType | APIError>(`/api/bill-type/${id}`, {
            name
        });

        return response.data;

    }

    static async deleteBillType(id : number) {

        const response = await this.axios.delete<BillType | APIError>(`/api/bill-type/${id}`);

        return response.data;

    }

}